File: sdks/rust/src/resources/payments.rs
//! Payment resource (`/v1/payments`).

use reqwest::Method;
use serde::{Deserialize, Serialize};

use crate::{Client, Entitlement, Payment, Result};

/// Accessor for payment lifecycle operations.
#[derive(Debug)]
pub struct Payments<'a> {
    client: &'a Client,
}

/// Body for recording a pending payment (`POST /v1/payments`).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RecordPayment {
    pub checkout_session_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tx_hash: Option<String>,
}

/// Body for confirming a payment (`POST /v1/payments/:id/confirm`).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ConfirmPayment {
    pub tx_hash: String,
    pub confirmations: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_confirmations: Option<i64>,
}

/// Result of confirming a payment: the payment plus any granted entitlements.
#[derive(Debug, Clone, Deserialize)]
pub struct ConfirmResult {
    pub payment: Payment,
    #[serde(default)]
    pub entitlements: Vec<Entitlement>,
}

impl<'a> Payments<'a> {
    pub(crate) fn new(client: &'a Client) -> Self {
        Payments { client }
    }

    /// Record a pending payment against a checkout session.
    pub async fn record(&self, body: &RecordPayment) -> Result<Payment> {
        self.client
            .request(Method::POST, "/v1/payments", Some(body))
            .await
    }

    /// Fetch a payment by id.
    pub async fn get(&self, id: &str) -> Result<Payment> {
        self.client
            .request_no_body(Method::GET, &format!("/v1/payments/{id}"))
            .await
    }

    /// Confirm a payment, completing its session and granting entitlements.
    pub async fn confirm(&self, id: &str, body: &ConfirmPayment) -> Result<ConfirmResult> {
        self.client
            .request(Method::POST, &format!("/v1/payments/{id}/confirm"), Some(body))
            .await
    }

    /// Fail a pending payment.
    pub async fn fail(&self, id: &str) -> Result<Payment> {
        self.client
            .request::<_, ()>(Method::POST, &format!("/v1/payments/{id}/fail"), None)
            .await
    }

    /// Refund a confirmed payment.
    pub async fn refund(&self, id: &str) -> Result<Payment> {
        self.client
            .request::<_, ()>(Method::POST, &format!("/v1/payments/{id}/refund"), None)
            .await
    }
}

File: sdks/python/settlekit/resources/payments.py
"""Payments resource — create, confirm, refund, and fetch payments."""

from __future__ import annotations

from typing import Any, Optional

from ._base import AsyncResource, SyncResource, clean_body


def _confirm_body(
    *, tx_hash: str, confirmations: int, min_confirmations: Optional[int]
) -> dict[str, Any]:
    return clean_body(
        {
            "txHash": tx_hash,
            "confirmations": confirmations,
            "minConfirmations": min_confirmations,
        }
    )


class Payments(SyncResource):
    """Synchronous payments namespace."""

    def create(self, *, checkout_session_id: str) -> Any:
        """Create a payment for a checkout session. ``POST /v1/payments``."""
        return self._request(
            "POST", "/v1/payments", body={"checkoutSessionId": checkout_session_id}
        )

    def confirm(
        self,
        payment_id: str,
        *,
        tx_hash: str,
        confirmations: int,
        min_confirmations: Optional[int] = None,
    ) -> Any:
        """Confirm an on-chain payment. ``POST /v1/payments/:id/confirm``."""
        body = _confirm_body(
            tx_hash=tx_hash,
            confirmations=confirmations,
            min_confirmations=min_confirmations,
        )
        return self._request("POST", f"/v1/payments/{payment_id}/confirm", body=body)

    def refund(self, payment_id: str) -> Any:
        """Refund a confirmed payment. ``POST /v1/payments/:id/refund``."""
        return self._request("POST", f"/v1/payments/{payment_id}/refund")

    def get(self, payment_id: str) -> Any:
        """Fetch a payment by id. ``GET /v1/payments/:id``."""
        return self._request("GET", f"/v1/payments/{payment_id}")


class AsyncPayments(AsyncResource):
    """Asynchronous payments namespace."""

    async def create(self, *, checkout_session_id: str) -> Any:
        """Create a payment for a checkout session. ``POST /v1/payments``."""
        return await self._request(
            "POST", "/v1/payments", body={"checkoutSessionId": checkout_session_id}
        )

    async def confirm(
        self,
        payment_id: str,
        *,
        tx_hash: str,
        confirmations: int,
        min_confirmations: Optional[int] = None,
    ) -> Any:
        """Confirm an on-chain payment. ``POST /v1/payments/:id/confirm``."""
        body = _confirm_body(
            tx_hash=tx_hash,
            confirmations=confirmations,
            min_confirmations=min_confirmations,
        )
        return await self._request(
            "POST", f"/v1/payments/{payment_id}/confirm", body=body
        )

    async def refund(self, payment_id: str) -> Any:
        """Refund a confirmed payment. ``POST /v1/payments/:id/refund``."""
        return await self._request("POST", f"/v1/payments/{payment_id}/refund")

    async def get(self, payment_id: str) -> Any:
        """Fetch a payment by id. ``GET /v1/payments/:id``."""
        return await self._request("GET", f"/v1/payments/{payment_id}")
